src/utils/simulation.js

```diff
@@ -0,0 +1,210 @@
+  }
+
+  /**
+   * Creates the initial population.
+   * People are randomly chosen to be initially sick and vaccinated according to given parameters.
+   * Infected people's ids are added to current_infected empty array.
+   * @return {array} pop - A list of Person objects.
+   */
+  
+  create_population() {
+    // create people objects
+    let pop = [];  
+    for (let i = 0; i < this.pop_size; i++) {
+      pop.push(new Person(i, false))
+    }
+
+    // calculate initial vaccination rate
+    const initial_vacc = Math.ceil(this.pop_size * this.vacc_percentage);
+    const randomPeopleIds = generateRandomnArray((this.initial_infected + initial_vacc),this.pop_size);
+
+    // infect people
+    for (let i = 0; i < this.initial_infected; i++) {
+      const random_person = randomPeopleIds[i]; 
+      pop[random_person].infection = this.virus;
+      this.current_infected.push(randomPeopleIds[i]);
+    }
+
+    // vaccinate people
+    for (let i = this.initial_infected; i < randomPeopleIds.length; i++) {
+      const random_person = randomPeopleIds[i];
+      pop[random_person].is_vaccinated = true; 
+    }
+
+    return pop;
+  }
+
+  /**
+   * The simulation should only end if the entire population is dead, everyone is vaccinated, or the virus died. 
+   * @return {boolean} - If simulation should continue. 
+   */
+  simulation_should_continue() {
+    // if no more infected people => end simulation since virus died
+    if (this.current_infected.length === 0) {
+      // console.log("virus ended");
+      return false;
+    }
+
+    for (let i = 0; i < this.population.length; i++) {
+      const person = this.population[i];
+      // continue if there's someone alive not vaccinate and sick
+      if (person.is_alive && !person.is_vaccinated) {
+        return true;
+      }
+    }
+    // all vaccinated/dead
+    return false;
+  }
+
+  /** 
+   * Runs the simulation until simulation_should_continue is false, which then console logs end message.
+   * If simulation continues, we add to our time step counter and 
+   * call the methods: time_step, kill_or_vaccinate_, infect_newly_infected.
+   */
+  run() {
+    let time_steps = 0;
+    while (this.simulation_should_continue()) {
+      // console.log("starting", time_steps + 1)
+      this.time_step();
+      time_steps += 1;
+      this.kill_or_vaccinate();
+      this.infect_newly_infected();
+    }
+    //console.log("alive sick still left", this.population.filter((person) => person.is_alive && !person.is_vaccinated));
+    //console.log("end pop", this.population);
+    console.log(`The simulation has ended after ${time_steps} turns.`);
+  }
+
+  /**
+   * @param {array} alive - People who are alive and not vaccinated.
+   * @param {number} count - The amount of people we want from the alive array.
+   * @returns {array} sample - Person objects who can get sick.
+   */
+  interaction_sample (alive, count) {
+    let sample = [];
+    let counter = 0; 
+
+    while (counter < count) {
```

Comment:
this can be turned into a for loop. something like this:`

```
for (; count > 0; count--) {
    doSomething();
}
```

Reply:
no need to create extra variable counter

Reply:
Changed to for loop

Reply:
do `count--` instead of `counter += 1` and turn the loop into a for loop

Reply:
Makes sense now, i updated it but still kept a while loop